### ISSUE TITLE
Classloader memory leak via JNI global references

### ISSUE THREAD

**Issue by SgtSilvio** [netty/netty#13480]
### Expected behavior

Netty (in this case via reactor-netty) can be used in conjunction with multiple classloaders, allowing classloaders to be properly unloaded if not referenced anymore. (in this case this is important for a gradle plugin)

### Actual behavior

When using `netty-resolver-dns-native-macos`, a classloader and all its classes can not be unloaded even if not referenced anymore as the class `io.netty.resolver.dns.macos.DnsResolver` is a GC root as a JNI global reference.

This class (and other required classes) are loaded via `NETTY_JNI_UTIL_LOAD_CLASS` (https://github.com/netty/netty/blob/79d3fa52dc0ee07ccc1d9c68648b57dc3bafc88e/resolver-dns-native-macos/src/main/c/netty_resolver_dns_macos.c#L186) which uses `NewGlobalRef` (https://github.com/netty/netty-jni-util/blob/8d820fba781882f8a7fe94efb7656d1413ff8bb8/src/main/c/netty_jni_util.h#L45).

Maybe using weak global references instead of global references of the classes is a solution.

Or is there any other solutions how to safely use netty with dynamic classloaders?

### Netty version

4.1.94.Final via reactor-netty:1.1.8

### JVM version (e.g. `java -version`)

Any

### OS version (e.g. `uname -a`)

MacOS

**Comment by normanmaurer** [comment]
I think the problem is that we don't call `NETTY_JNI_UTIL_UNLOAD_CLASS ` for it ... Let me fix this

**Comment by normanmaurer** [comment]
@SgtSilvio see https://github.com/netty/netty/pull/13481

**Comment by SgtSilvio** [comment]
Thanks for the quick answer and nice catch. It fixes definitely an issue, but are you sure it is enough to allow the GC to collect the classloader? I thought the unload of the native library will only happen once the classloader is ready for GC, but I might be mistaken (as my JNI knowledge is very limited).

**Comment by normanmaurer** [comment]
@SgtSilvio I think it should work, that said I might also be wrong :)

**Comment by SgtSilvio** [comment]
ok, I want to try this out. Is there an easy way to get a snapshot of `netty-resolver-dns-native-macos`. Or which steps do I need to take to build it myself? A simple maven package does not seem to be enough.

**Comment by normanmaurer** [comment]
@SgtSilvio this should work: 

```
./mvnw -pl transport-native-unix-common,resolver-dns-native-macos,resolver-dns-classes-macos clean package
```

**Comment by SgtSilvio** [comment]
getting the following error:
> [ERROR] Failed to execute goal org.fusesource.hawtjni:hawtjni-maven-plugin:1.18:build (build-native-lib) on project netty-resolver-dns-native-macos: build failed: org.apache.maven.plugin.MojoExecutionException: Requested download does not exist.: The following artifacts could not be resolved: io.netty:netty-resolver-dns-native-macos:zip:native-src:4.1.95.Final-SNAPSHOT (absent)

**Comment by SgtSilvio** [comment]
@normanmaurer I tried it out and unfortunately it seems I was right.
As long as the classloader is still referenced - in this case by the `io.netty.resolver.dns.macos.DnsResolver` class that was loaded by the classloader itself and is registered as a JNI global reference - the native library will not be unloaded and so the JNI global reference will not be freed.
This means that the calls to `NETTY_JNI_UTIL_UNLOAD_CLASS` inside a `JNI_OnUnLoad` class are basically useless as they will never be executed.
This basically means that netty can not really be used in a JVM that uses some sort of plugin system that might want to create and release new classloaders dynamically (for example gradle) which is very unfortunate.

I guess I should open a new issue, right?

**Comment by normanmaurer** [comment]
Yes … that said I am not really sure what to do about atm Am 12.07.2023 um 15:53 schrieb Silvio Giebl ***@***.***>:﻿
@normanmaurer I tried it out and unfortunately it seems I was right.
As long as the classloader is still referenced - in this case by the io.netty.resolver.dns.macos.DnsResolver class that was loaded by the classloader itself and is registered as a JNI global reference - the native library will not be unloaded and so the JNI global reference will not be freed.
This means that the calls to NETTY_JNI_UTIL_UNLOAD_CLASS inside a JNI_OnUnLoad class are basically useless as they will never be executed.
This basically means that netty can not really be used in a JVM that uses some sort of plugin system that might want to create and release new classloaders dynamically (for example gradle) which is very unfortunate.
I guess I should open a new issue, right?

**Comment by SgtSilvio** [comment]
> Maybe using weak global references instead of global references of the classes is a solution.

What do you think about this? It seems like using a weak global reference might be right in general for references to classes. Because, as long as the classloader is referenced from at least one place in the JVM, the weak JNI reference will stay there as well. (as of my current understanding)

**Comment by normanmaurer** [comment]
This "might" work. Can you patch it and test it ? If the maven command which I shared with you did not work, please show the full output. Most likely you miss the tools that are needed to compile native stuff:

https://netty.io/wiki/native-transports.html#wiki-h2-4

**Comment by SgtSilvio** [comment]
>  If the maven command which I shared with you did not work, please show the full output. Most likely you miss the tools that are needed to compile native stuff

Already worked for me, issue was a missing tool.

**Comment by normanmaurer** [comment]
ok cool... then please try the change and let me know how it goes.

**Comment by SgtSilvio** [comment]
@normanmaurer First success. Using a weak global reference is working as expected.
Here are the quick changes that I made for testing this: https://github.com/SgtSilvio/netty/commit/4b690f04c9dd2a60b31aa38cf981d11a16eb4778

**Comment by normanmaurer** [comment]
From here:
https://www.ibm.com/docs/en/ztpf/1.1.0.15?topic=collection-jni-weak-reference

```
JNI weak references provide the same capability as that of WeakReference objects, but the processing is very different. A JNI routine can create a JNI Weak reference to an object and later delete that reference. The Garbage Collector clears any weak reference where the referent is unmarked, but no equivalent of the queuing mechanism exists.

Failure to delete a JNI Weak reference causes a memory leak in the table and performance problems. This also applies to JNI global references. The processing of JNI weak references is handled last in the reference handling process. The result is that a JNI weak reference can exist for an object that has already been finalized and had a phantom reference queued and processed.
```

So this would leak memory, no ?

**Comment by SgtSilvio** [comment]
> So this would leak memory, no ?

If the reference is not deleted, then it creates a memory leak. This applies to global JNI references and weak global JNI references.
Of course, the reference itself is still deleted, just a `DeleteWeakGlobalRef` at the same place as the current `DeleteGlobalRef`
(See `DeleteWeakGlobalRef` in my changes https://github.com/SgtSilvio/netty/commit/4b690f04c9dd2a60b31aa38cf981d11a16eb4778#diff-c8559d8c44f0e5f95ae8239541fa8632fa502f43b4cef062b69236340e65c2b7R52 and `DeleteGlobalRef` in the current code
https://github.com/netty/netty-jni-util/blob/main/src/main/c/netty_jni_util.h#L55C27-L55C27)

**Comment by SgtSilvio** [comment]
Currently:
1. Classloader references JNI native library (the native library will not be unloaded until the classloader is garbage collected)
2. JNI global reference from the native library **STRONGLY** references `io.netty.resolver.dns.macos.DnsResolver` class
3. `io.netty.resolver.dns.macos.DnsResolver` class strongly references the classloader
=> we have circular strong references between classloader and native library.
=> the native library will never be unloaded as there will always be a strong reference to the classloader from the JNI global reference, and this reference will not be deleted because it would only be done when the native library is unloaded.

When using a weak global reference for the `io.netty.resolver.dns.macos.DnsResolver` class:
1. Classloader references JNI native library (the native library will not be unloaded until the classloader is garbage collected)
2. JNI weak global reference from the native library **WEEKLY** references `io.netty.resolver.dns.macos.DnsResolver` class 
3. `io.netty.resolver.dns.macos.DnsResolver` class strongly references the classloader
=> as the weak reference does not count towards the reference count of an object, the classloader can be garbage collected once no reference to the classloader exists anymore from the Java code.
=> once the classloader is garbage collected, the native library is unloaded, `JNI_OnUnload` gets called and the weak global reference (basically the wrapper object, the referent is already garbage collected) will be deleted. 
=> no memory leak

**Comment by normanmaurer** [comment]
Let me do some experiments

**Comment by normanmaurer** [comment]
@SgtSilvio can you provide some sort of self-containing reproducer ? I am not convinced yet that the weak reference is the best way how to fix this

**Comment by SgtSilvio** [comment]
@normanmaurer sure I can try to assemble something, but I need to know what you want the reproducer to show/reproduce.
What exactly are you unsure about, is it:
1. You don't think that this actually fixes the issue (that the classloader can be garbage collected when using JNI weak global references) or
2. You think there must be a better way?

For the first point, I want to mention that GC of classloaders seems to be very lazy and I see it delayed by several minutes, but heapdumps show that the classloader actually has no GC root, and this really fixes the issue.
For the second point, what are your concerns why JNI weak global references are not "good"?

**Comment by normanmaurer** [comment]
I am not sure if this is the best way to fix it... Using a weak-reference feels more like a workaround.

**Comment by SgtSilvio** [comment]
I understand, whenever I see a weak reference it also looks suspicious to me first.
But in this case it feels correct to me:
The application has a strong reference on classloader with netty.
The classloader has a strong reference on the native library.
Whenever the application decides to not need the classloader anymore, everything (metaspace and native state) should be able to be unloaded.
The native library needs a reference to the class for creating objects etc. only as long as the classloader and its classes are still in use.

Also, if you think it is bad to rely on GC for unloading, here I don't think it is a problem. Unloading is an infrequent operation and it is fine to happen eventually (and not immediately) as the goal is to avoid a long-running process to run OOM in the long run. 

What would be alternatives? I can only think about manually calling a native `unload` method, but the user does not know when this method can be called. In my case I have no control over when a classloader is no longer in use anymore.
Or we do not create global variables on classes at all, but this means that the code always needs to search for the class over and over again which is very likely bad for performance (as far as I understand JNI).

**Comment by SgtSilvio** [comment]
Please don't get me wrong, I am not trying to talk you into using weak references as the final solution, but I do not see alternatives as of now.
And I can still try to assemble a reproducer. The producer will very likely require to take a heap dump to verify if the classloader is still referenced by a GC root. Is that ok for you?

**Comment by normanmaurer** [comment]
@SgtSilvio yes taking a heap dump is fine...I am also not sure if there is any other way to solve it but I would like to explore this first before making any changes.

**Comment by SgtSilvio** [comment]
@normanmaurer here is the reproducer code: https://github.com/SgtSilvio/netty-reproducer-13480

**Comment by SgtSilvio** [comment]
@normanmaurer is the reproducer what you have expected or do you need more information?
What can I do to move this forward?

**Comment by normanmaurer** [comment]
@SgtSilvio I just started to look into this again... So I think we would need to load all classes that are not part of the JDK as weak references, correct ?

**Comment by SgtSilvio** [comment]
Yes, JDK classes/any classes in the boot classloader are shared and are not a problem, because they don't reference a classloader that can be unloaded.

**Comment by normanmaurer** [comment]
@SgtSilvio PTAL https://github.com/netty/netty/pull/13584

**Comment by agilob** [comment]
Sorry to bring this topic to life again, but how can I tell if I'm affected by this leak? I've been chasing a native memory leak for a few weeks now, now I'm looking into `netty-transport-native-epoll-4.1.97`. I have access to a few heapdump, jfr recordings, and jemalloc snapshots.

>For the first point, I want to mention that GC of classloaders seems to be very lazy and I see it delayed by several minutes, but heapdumps show that the classloader actually has no GC root, and this really fixes the issue.

Would I be looking for classloader objects that simply have no GC root?

**Comment by SgtSilvio** [comment]
@agilob this issue has been fixed since 4.1.98, so it is possible that you hit it as you are still on 4.1.97.
But it depends on what your application does. This issue only happens if you create multiple classloaders that are then later unloaded.

> Would I be looking for classloader objects that simply have no GC root?

No, you should see a JNI global reference as a GC root. The following screenshot shows how this can look like in VisualVM:

<img width="758" alt="image" src="https://github.com/netty/netty/assets/28134127/bf8a4c06-d3bf-4111-aa6a-cbe262cd97d1">
